src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -138,6 +129,49 @@ ss::future<walk_result> recursive_directory_walker::walk(
                 throw;
             }
         }
+    };
+
+    // Listing directories involves blocking I/O which in Seastar is serviced by
+    // a dedicated thread pool and workqueue. When listing directories with
+    // large number of sub-directories this leads to a large queue of work items
+    // for each of which we have to incur the cost of task switching. Even for a
+    // lightly loaded reactor, it can take up to 1ms for a full round-trip. The
+    // round-trip time is ~100x the syscall duration. With 2 level directory
+    // structure, 100K files, 2 getdents64 calls per directory this adds up to
+    // 400K syscalls and 6m of wall time.
+    //
+    // By running the directory listing in parallel we also parallelize the
+    // round-trip overhead. This leads to a significant speedup in the
+    // directory listing phase.
+    //
+    // Limit the number of concurrent directory reads to avoid running out of
+    // file descriptors.
+    //
+    // Empirical testing shows that this value is a good balance between
+    // performance and resource usage.
+    const size_t max_concurrency = 1000;
+
+    std::vector<ss::sstring> targets;
+    targets.reserve(max_concurrency);
+
+    while (!state.empty()) {
+        targets.clear();
+
+        auto concurrency = std::min(
+          state.dirlist.size(), size_t(max_concurrency));
+
+        for (size_t i = 0; i < concurrency; ++i) {
+            auto target = state.pop();
+            vassert(
+              std::string_view(target).starts_with(start_dir),
+              "Looking at directory {}, which is outside of initial dir "
+              "{}.",
+              target,
+              start_dir);
+            targets.push_back(std::move(target));
+        }
+
+        co_await ss::parallel_for_each(targets, process_directory);
```

Comment:
You can use max_concurrent_for_each instead and get rid of outer while-loop.

Reply:
The outer while loop is still required because the `state` is changed as we process the directory tree. Even though it ends up being empty at some point, the `process_directory` might add new entries to it.